tracing-opentelemetry/src/span_ext.rs

```diff
@@ -51,47 +53,95 @@ pub trait OpenTelemetrySpanExt {
             if let Some(get_context) = subscriber.downcast_ref::<WithContext>() {
-                get_context.with_context(subscriber, id, move |builder| {
-                    builder.parent_context = parent_context.take()
+                get_context.with_context(subscriber, id, move |builder, _sampler| {
+                    builder.parent_context = parent_context.remote_span_context().cloned()
                 });
             }
         });
     }
 
-    fn context(&self) -> api::SpanContext {
+    fn context(&self) -> api::Context {
         let mut span_context = None;
         self.with_subscriber(|(id, subscriber)| {
             if let Some(get_context) = subscriber.downcast_ref::<WithContext>() {
-                get_context.with_context(subscriber, id, |builder| {
-                    span_context = Some(build_context(builder));
+                get_context.with_context(subscriber, id, |builder, sampler| {
+                    span_context = Some(build_span_context(builder, sampler));
                 })
             }
         });
 
-        span_context.unwrap_or_else(api::SpanContext::empty_context)
+        let compat_span = CompatSpan(span_context.unwrap_or_else(api::SpanContext::empty_context));
+        api::Context::current_with_span(compat_span)
+    }
+}
+
+/// A compatibility wrapper for an injectable OpenTelemetry span context.
+#[derive(Debug)]
+struct CompatSpan(api::SpanContext);
+impl api::Span for CompatSpan {
+    fn add_event_with_timestamp(
+        &self,
+        _name: String,
+        _timestamp: std::time::SystemTime,
+        _attributes: Vec<api::KeyValue>,
+    ) {
+        // OpenTelemetry and tracing APIs cannot be mixed, use `tracing::event!`
+        // macro instead.
+    }
+
+    /// This method is used by OpenTelemetry propagators to inject span context
+    /// information into [`Carrier`]s.
+    ///
+    /// [`Carrier`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/context/propagation/trait.Carrier.html
+    fn span_context(&self) -> api::SpanContext {
+        self.0.clone()
+    }
+
+    fn is_recording(&self) -> bool {
+        // cannot record via OpenTelemetry API when using extracted span in tracing
+        false
+    }
+
+    fn set_attribute(&self, _attribute: api::KeyValue) {
+        // OpenTelemetry and tracing APIs cannot be mixed, use `tracing::span!`
+        // macro or `span.record()` instead.
+    }
+
+    fn set_status(&self, _code: api::StatusCode, _message: String) {
+        // OpenTelemetry and tracing apis cannot be mixed, use `tracing::span!`
+        // macro or `span.record()` instead.
+    }
+
+    fn update_name(&self, _new_name: String) {
+        // OpenTelemetry and tracing APIs cannot be mixed, span names are not
+        // mutable.
+    }
+
+    fn end(&self) {
+        // OpenTelemetry and tracing APIs cannot be mixed, span end times are set
+        // when the underlying tracing span closes.
```

Comment:
Should we document that these functions cannot be used while using `tracing`? Perhaps they should panic if we're in debug mode, so that it's clear to the user that they shouldn't be used, but they don't crash the app in production?

Reply:
Good idea.

Reply:
switched to `debug_assert!(false)`, is there a better way to accomplish the same thing?

Reply:
```rust
#[cfg(debug_assertions)]
panic!("OpenTelemetry `Span::end` API cannot be used when blah blah blah et cetera");
```
should work?

Reply:
Thanks! clearer 👍